### PR TITLE
Fix assertIngestDocument wrongfully passing

### LINKED ISSUE #28492: assertIngestDocument() is weaker than it appears
It seems that `IngestDocumentMatcher#assertIngestDocument()` does not look at the content of its arguments too carefully. The following test does not pass:

```diff
diff --git a/server/src/test/java/org/elasticsearch/ingest/IngestDocumentTests.java b/server/src/test/java/org/elasticsearch/ingest/IngestDocumentTests.java
index 9df2a38c6f..0ab5d79457 100644
--- a/server/src/test/java/org/elasticsearch/ingest/IngestDocumentTests.java
+++ b/server/src/test/java/org/elasticsearch/ingest/IngestDocumentTests.java
@@ -1021,4 +1021,13 @@ public class IngestDocumentTests extends ESTestCase {
         }
     }
 
+    public void testEqualityAssertion() {
+        Map<String, Object> sourceAndMetadata1 = new HashMap<>();
+        sourceAndMetadata1.put("foo", "bar");
+        IngestDocument document1 = new IngestDocument(sourceAndMetadata1, new HashMap<>());
+
+        IngestDocument document2 = new IngestDocument(new HashMap<>(), new HashMap<>());
+
+        expectThrows(AssertionError.class, () -> assertIngestDocument(document1, document2));
+    }
 }

```

Found during the course of https://github.com/elastic/elasticsearch/pull/28476

### PR DESCRIPTION
**Pull request by original-brownbear**
* Previously docA being subset of docB passed because iteration was over docA's keys only
* Scalars in nested fields were not compared in all cases
* Assertion errors were hard to interpret (message wasn't correct since it only mentioned the class type)
* In cases where two paths contained different types a ClassCastException was thrown instead of an AssertionError
* Adjusted a few tests that now failed (but were generally correct in behaviour imo except for the assertion)
* Fixes #28492

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
Jenkins test this

File: test/framework/src/test/java/org/elasticsearch/ingest/IngestDocumentMatcherTests.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.elasticsearch.ingest.IngestDocumentMatcher.assertIngestDocument;
+
+public class IngestDocumentMatcherTests extends ESTestCase {
+
+    public void testAssertIngestDocumentFailsOnDifferentMapData() {
```
Comment (rjernst) [inline test/framework/src/test/java/org/elasticsearch/ingest/IngestDocumentMatcherTests.java]: These names would be a lot easier to read without the redundant info. Eg `testDifferentMapData()`

**Review by rjernst — APPROVED** [review]
LGTM

File: test/framework/src/test/java/org/elasticsearch/ingest/IngestDocumentMatcherTests.java (hunk 1 above)
Comment (original-brownbear) [reply test/framework/src/test/java/org/elasticsearch/ingest/IngestDocumentMatcherTests.java]: +1, will rename before merging, thanks! :)

**Review by martijnvg — APPROVED** [review]
👍

**Pull request merged**